src/main/java/com/rsr5/gocd/riemann_notifier/GoNotificationPlugin.java

```diff
@@ -138,23 +144,80 @@ private GoPluginApiResponse handleNotificationsInterestedIn() {
         return renderJSON(SUCCESS_RESPONSE_CODE, response);
     }
 
-    private String service(JsonObject json) {
-        JsonObject pipelineObject, stageObject;
-        pipelineObject = (JsonObject) json.get("pipeline");
-        stageObject = (JsonObject) pipelineObject.get("stage");
+    private String safeGetAsString(JsonObject json, String property) {
+        JsonElement element = json.get(property);
+        if (element != null && element.isJsonPrimitive()) {
+            return element.getAsString();
+        }
+        return null;
+    }
 
-        String pipeline = pipelineObject.get("name").getAsString();
-        String stage = stageObject.get("name").getAsString();
+    final class StageData {
```

Comment:
Minor: do these classes close over anything from their parents? It doesn't look like it, in which case these should ideally also have a `static` modifier. `final` means you can't inherit from the class, which is also fine here since we don't expect any subclasses.